db/view/view.cc

```diff
@@ -1109,102 +1127,65 @@ future<> mutate_MV(
+        // First, find the local endpoint and ensure that if it exists,
+        // it will be the paired endpoint. That way, all endpoints in the
+        // pending_endpoints list are guaranteed to be remote.
+        auto my_address = utils::fb_utilities::get_broadcast_address();
+        auto pending_it = std::find(pending_endpoints.begin(), pending_endpoints.end(), my_address);
+        if (pending_it != pending_endpoints.end()) {
+            if (!paired_endpoint) {
+                paired_endpoint = *pending_it;
+                pending_endpoints.erase(pending_it);
             } else {
-                vlogger.debug("Sending view update to endpoint {}, with pending endpoints = {}", *paired_endpoint, pending_endpoints);
-                // Note we don't wait for the asynchronous operation to complete
-                // without a batchlog, and without checking for success.
-                // When the ownership of the view partition is being moved to a
-                // new node (or nodes), listed in pending_enpoints, we also need
-                // to send the update there. Currently, we do this from *each* of
-                // the base replicas, but this is probably excessive - see
-                // See https://issues.apache.org/jira/browse/CASSANDRA-14262/
-                tracing::trace(tr_state, "Sending view update for {}.{} to {}, with pending endpoints = {}; base token = {}; view token = {}",
-                        mut.s->ks_name(), mut.s->cf_name(), *paired_endpoint, pending_endpoints, base_token, view_token);
-                future<> view_update = service::get_local_storage_proxy().send_to_endpoint(
-                        std::move(mut),
-                        *paired_endpoint,
-                        std::move(pending_endpoints),
-                        db::write_type::VIEW,
-                        std::move(tr_state),
-                        stats,
-                        allow_hints).then_wrapped(
-                                [paired_endpoint,
-                                 is_endpoint_local,
-                                 updates_pushed_remote,
-                                 maybe_account_failure = std::move(maybe_account_failure)] (future<>&& f) mutable {
-                    return maybe_account_failure(std::move(f), std::move(*paired_endpoint), is_endpoint_local, updates_pushed_remote);
-                });
-                if (wait_for_all) {
-                    fs->push_back(std::move(view_update));
+                // Remove the duplicated entry
+                if (*paired_endpoint == *pending_it) {
+                    pending_endpoints.erase(pending_it);
                 } else {
-                    // The update is sent to background in order to preserve availability,
-                    // its parallelism is limited by view_update_concurrency_semaphore
-                    (void)view_update;
+                    std::swap(*paired_endpoint, *pending_it);
```

Comment:
Well, yeah, I wrote it in a funky way precisely because `send_to_endpoint()` function wants to distinguish one endpoint as `target`. But, what I discovered lately, is that the same `send_to_endpoint()` function later aggregates this target back to a single list along with all the pending endpoints... I'll do this patch the other way round - start with fixing `send_to_endpoint` to better fit our needs, and then use it in `mutate_MV`. This flavor of `send_to_endpoint` is only used by materialized view code, so there's no problem in changing its definition. Then, once `send_to_endpoint` will simply accept a list of endpoints to send the update to, `mutate_MV` will naturally become simpler as well.

Reply:
Hm, no, the notion of `target` is actually used, I overlooked it. Well, alright, I'll still rewrite this pull request so that I don't reuse `paired_endpoint` in a weird way.

Reply:
If the notion of "target" has a special meaning (what is it?), then is it actually ok that this code exchanges the paired endpoint and pending endpoint just to make sure the latter is the remote one?

Reply:
The special meaning actually belongs to `pending_endpoints` and it's quite lame - pretty much just to record their size in the response handler... so that's why we split the primary target and the pending endpoints. Refactoring these is more dangerous however, since the same code is also used in other modules, e.g. hints.

To sum up - I gave a try at the algorithm you sketched above, since it's quite similar to what happens now, but I don't like the notion of putting the paired endpoint into the list of remote endpoints, just to pop it back a moment later in order to provide it as a "target". So while the current code looks funky, it saves us a push+pop on a vector, which may actually involve an allocation. Perhaps I'll just clean the current solution up, change some names and add comments?

Also, there's still a matter of me not being able to run this dtest locally due to unrelated repair errors. I need to figure it out before I mark this patch as ready, since I'd like to ensure that this test suceeds.

Reply:
Actually, just renaming `s/paired_endpoint/target_endpoint` and `s/pending_endpoints/remote_endpoints` makes the code way more clear. I'll clean it up more and post results here later today.